src/lib.rs

```diff
@@ -0,0 +1,178 @@
+    };
+
+    tracing::debug!("Looking for index file at: {}", index_path.display());
+
+    match std::fs::read(&index_path) {
+        Ok(content) => {
+            tracing::info!(
+                "Successfully served index for {}, {} bytes",
+                crate_name,
+                content.len()
+            );
+            let mut response = Response::new(axum::body::Body::from(content));
+            response.headers_mut().insert(
+                axum::http::header::CONTENT_TYPE,
+                "text/plain".parse().unwrap(),
+            );
+            Ok(response)
+        }
+        Err(e) => {
+            tracing::warn!("Failed to read index file for {}: {}", crate_name, e);
+            Err(StatusCode::NOT_FOUND)
+        }
+    }
+}
+
+async fn serve_crate_file(
+    axum::extract::State((registry_path, _)): axum::extract::State<(PathBuf, String)>,
+    AxumPath(filename): AxumPath<String>,
+) -> Result<Response, StatusCode> {
+    if filename.ends_with(".crate") {
+        tracing::info!("Serving crate file: {}", filename);
+        let crate_path = registry_path.join(&filename);
+
+        tracing::debug!("Looking for crate file at: {}", crate_path.display());
+
+        match std::fs::read(&crate_path) {
+            Ok(content) => {
+                tracing::info!(
+                    "Successfully served crate file {}, {} bytes",
+                    filename,
+                    content.len()
+                );
+                let mut response = Response::new(axum::body::Body::from(content));
+                response.headers_mut().insert(
+                    axum::http::header::CONTENT_TYPE,
+                    "application/octet-stream".parse().unwrap(),
+                );
+                Ok(response)
+            }
+            Err(e) => {
+                tracing::warn!("Failed to read crate file {}: {}", filename, e);
+                Err(StatusCode::NOT_FOUND)
+            }
+        }
+    } else {
+        Err(StatusCode::NOT_FOUND)
+    }
+}
+
+async fn serve_file(
+    axum::extract::State((registry_path, _)): axum::extract::State<(PathBuf, String)>,
+    uri: axum::http::Uri,
+) -> Result<Response, StatusCode> {
+    let file_path = uri.path().trim_start_matches('/');
+    tracing::info!("Fallback file request for: {}", file_path);
+    let full_path = registry_path.join(file_path);
+
+    tracing::debug!("Looking for file at: {}", full_path.display());
+
+    if !full_path.starts_with(&registry_path) {
+        return Err(StatusCode::FORBIDDEN);
+    }
+
+    match std::fs::read(&full_path) {
+        Ok(content) => {
+            let content_len = content.len();
+            let mut response = Response::new(axum::body::Body::from(content));
+
+            if let Some(ext) = full_path.extension().and_then(|e| e.to_str()) {
+                let content_type = match ext {
+                    "json" => "application/json",
+                    "tar" | "gz" => "application/gzip",
```

Comment:
Nitpicking, but I think you shouldn't bind tar to `application/gzip`,  a tarball is not necessarily compressed or can be compressed using another compression algo